### PR TITLE
db/hints: Cancel draining when stopping node

### PR DESCRIPTION
**Pull request by dawmd**
Draining hints may occur in one of the two scenarios:

* a node leaves the cluster and the local node drains all of the hints saved for that node,
* the local node is being decommissioned.

Draining may take some time and the hint manager won't stop until it finishes. It's not a problem when decommissioning a node, especially because we want the cluster to retain the data stored in the hints. However, it may become a problem when the local node started draining hints saved for another node and now it's being shut down.

There are two reasons for that:

* Generally, in situations like that, we'd like to be able to shut down nodes as fast as possible. The data stored in the hints won't disappear from the cluster yet since we can restart the local node.
* Draining hints may introduce flakiness in tests. Replaying hints doesn't have the highest priority and it's reflected in the scheduling groups we use as well as the explicitly enforced throughput. If there are a large number of hints to be replayed, it might affect our tests. It's already happened, see: scylladb/scylladb#21949.

To solve those problems, we change the semantics of draining. It will behave as before when the local node is being decommissioned. However, when the local node is only being stopped, we will immediately cancel all ongoing draining processes and stop the hint manager. To amend for that, when we start a node and it initializes a hint endpoint manager corresponding to a node that's already left the cluster, we will begin the draining process of that endpoint manager right away.

That should ensure all data is retained, while possibly speeding up the shutdown process.

There's a small trade-off to it, though. If we stop a node, we can then remove it. It won't have a chance to replay hints it might've before these changes, but that's an edge case. We expect this commit to bring more benefit than harm.

We also provide tests verifying that the implementation works as intended.

Fixes scylladb/scylladb#21949

Backport: not needed. This is an enhancement.

### CODE REVIEW

File: test/topology_custom/test_hints.py
Comment (dawmd) [inline test/topology_custom/test_hints.py:1]: I guess putting the tests in this file might need some explanation. I didn't write them in the topology suite because the cluster is going to be marked dirty anyway. Besides, I think it's better to keep the hint tests in a single file than scatter them throughout the project.

**Comment by dawmd** [comment]
This PR is a draft because I observed some errors related to draining a node that's already left at startup:

```
ERROR 2025-02-11 20:09:32,844 [shard 0:strm] hints_manager - Exception when draining b936c6ff-1f2c-4789-a951-a1b1905b7933: std::filesystem::__cxx11::filesystem_error (error system:39, filesystem error: remove failed: Directory not empty [/home/dawidmedrek/src/scylladb/testlog/dev/scylla-1/hints/0/b936c6ff-1f2c-4789-a951-a1b1905b7933])
```

I'll figure it out.

**Comment by dawmd** [comment]
Version 1:
* Moved the caller of hint draining to main.cc. See my review comment for more information.
* Added more comments in the tests. Adjusted the timeouts in them.
* Tests pass. There are no more errors in the logs.

File: main.cc [hunk 1]
```diff
@@ -2263,6 +2263,24 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             supervisor::notify("allow replaying hints");
             proxy.invoke_on_all(&service::storage_proxy::allow_replaying_hints).get();
 
+            // Note: This is here because we can only start draining after we allow for replaying hints.
```
Comment (dawmd) [inline main.cc:2266]: I'm very sad to make main.cc the caller of this. I was thinking about some other places (resource manager, storage proxy, hint manager, etc.) that could do it instead, but I think the implementation would end up being convoluted or prone to errors/misunderstandings in the future.

The crucial thing here is that we must have enabled replaying hints by the point we try to drain the hint endpoint managers. Not satisfying that condition would lead drain to finish almost immediately and possibly without sending a single hint out. As a result, we'd observe errors in the logs and not achieve anything. On the other hand, we want to start draining as soon as we're able to to avoid more more complex situations related to the exposed API (notice that we start it right after these calls): for example, the user may decide to change the host filter, which would affect which endpoint manager are active and can send hints. That could get in our way. Putting this call in main helps us prevent us—we don't have a guarantee that the draining will finish before we expose the API to the user, but we do have a guarantee that all the drain calls will have been queued up by that point. That should be enough caution. If there's an edge case I didn't think of, it should be quite rare, but it shouldn't have any critical consequences for the user, i.e. no error should be logged; only the draining process may be interrupted, but that seems to be an acceptable scenario.

**Comment by piodul** [comment]
> There's a small trade-off to it, though. If we stop a node, we can then remove it. It won't have a chance to replay hints it might've before these changes, but that's an edge case. We expect this commit to bring more benefit than harm.

You can say the same about unsent hints that this node has towards other - live - nodes as well. In general, if you have a cooperative node (i.e. it didn't crash and become unrecoverable), you will rather want to use `nodetool decomission`. AFAIK stopping the node afterwards will wait for those hints to be sent out (perhaps `decommission` should wait for this to finish instead). Therefore, I think this is completely fine.

**Comment by dawmd** [comment]
Version 2.2:
* Rebased to resolve the merge conflict.

**Comment by dawmd** [comment]
> > There's a small trade-off to it, though. If we stop a node, we can then remove it. It won't have a chance to replay hints it might've before these changes, but that's an edge case. We expect this commit to bring more benefit than harm.
> 
> You can say the same about unsent hints that this node has towards other - live - nodes as well. In general, if you have a cooperative node (i.e. it didn't crash and become unrecoverable), you will rather want to use `nodetool decomission`. AFAIK stopping the node afterwards will wait for those hints to be sent out (perhaps `decommission` should wait for this to finish instead). Therefore, I think this is completely fine.

That makes sense. I left the original wording of the commit message just to highlight that.

**Comment by mykaul** [comment]
What's the status of this issue? It might be the cause of multiple CI issues (unsure).

**Comment by dawmd** [comment]
> What's the status of this issue? It might be the cause of multiple CI issues (unsure).

@mykaul, the PR still needs to get approved. @piodul has already looked at it, but we had a few slips with CI/rebasing as well as some other important patches to be reviewed that took most of our team's attention.

@gusev-p, @denesb, adding you as reviewers because you're probably the most familiar with hinted handoff outside our team. If you can spare a bit of your time, please take a look. If not, that's fine.

**Comment by mykaul** [comment]
@scylladb/scylla-maint - please review this for inclusion. My hope is that merging this will reduce significantly the number of CI failures we see in random tests, many originate from this issue.
